src/main/java/org/craftercms/engine/service/context/SiteContext.java

```diff
@@ -138,17 +138,23 @@ public static SiteContext getCurrent() {
     /**
      * Sets the context for the current thread.
      */
-    public static void setCurrent(SiteContext current) {
-        logger.debug("Getting access lock for context {}", current);
-        current.accessLock.lock();
+    public static void setCurrent(SiteContext siteContext) {
+        SiteContext current = threadLocal.get();
+        if (current != null) {
+            logger.debug("Overwriting previous site context {}", current);
+            release(current);
+        }
 
-        if (current.scriptSandbox != null) {
-            current.scriptSandbox.register();
+        logger.debug("Getting access lock for context {}", siteContext);
+        siteContext.accessLock.lock();
+
```

Comment:
this method only handles the locking, unlocking is handled by the `clear()` method, https://github.com/craftercms/engine/blob/support/3.1.x/src/main/java/org/craftercms/engine/servlet/filter/SiteContextResolvingFilter.java#L87

Reply:
Yes I understand, I'm just saying if the code:

```
        if (siteContext.scriptSandbox != null) {
            siteContext.scriptSandbox.register();
        }

        threadLocal.set(current);
        threadLocal.set(siteContext);

        MDC.put(SITE_NAME_MDC_KEY, current.getSiteName());
        MDC.put(SITE_NAME_MDC_KEY, siteContext.getSiteName());
```

Fails with a weird `RuntimeException` or `Error`, you will still be leaking a lock. It's better to be safe and make sure that any code that is inside that method has a `try...catch` with a `unlock` in the catch so that we remove any possibility of leaking a lock.